YJDH-251 | KS-Backend: Enforce youth application handler ADFS login

karisal-anders opened the pull request:
## Description :sparkles:

### KS-Backend: Enforce youth application handler ADFS login 

Enforce youth application handler ADFS login by adding
@enforce_handler_view_adfs_login decorator to:
 - YouthApplicationViewSet.process
   - Handler's processing email contains a link to this action
 - YouthApplicationViewSet.detail
   - Used by frontend to get data for processing a youth application

Revamp permission handling in YouthApplicationViewSet.

Add tests:
 - test_youth_applications_forbidden_modification_methods
   - Test that forbidden methods in YouthApplicationViewSet return
     HTTP status 403 Forbidden which can also seen in the audit log
 - test_youth_application_post_valid_audit_log
   - Test that the posting of a valid youth application can be seen in
     the audit log

Extend test with more parametrizations and audit log entry testing:
 - test_youth_applications_list

Extend tests with more parametrizations and redirection target testing:
 - test_youth_applications_process_valid_pk
 - test_youth_applications_process_unused_pk
 - test_youth_applications_detail_valid_pk
 - test_youth_applications_detail_unused_pk

Shared code changes:
 - Change HelsinkiOAuth2CallbackView.get:
   - Add support for USE_ORIGINAL_REDIRECT_URL session key
     which in case of a redirect which is not a multi-factor
     authentication redirect redirects to the original redirect URL.
     Otherwise works as before i.e. redirects to
     settings.ADFS_LOGIN_REDIRECT_URL.
   - Re-enable multi-factor authentication redirect by trying to detect
     it and letting it pass through unchanged. This was previously
     suppressed, the user was instead redirected to
     settings.ADFS_LOGIN_REDIRECT_URL. NOTE: This code path is untested.
 - Add test test_adfs_callback_original_redirect
   - Test that HelsinkiOAuth2CallbackView callback function redirects
     to the original redirect URL if session's dictionary item
     USE_ORIGINAL_REDIRECT_URL is True and that the dictionary
     item gets removed.
 - Add get_default_test_host function
 - Add permit_view_if_permissions_or_redirect decorator
   - This is a decorator for permitting a view if all permissions pass,
     otherwise redirecting to login URL with optional check (e.g. to
     remove redirection loops) and side effect (e.g. set session
     variables) function calls before redirection. If not all
     permissions pass nor redirection is taken then returns 403
     Forbidden.
 - Add redirect_to_login_using_request function
   - Redirects to login page similarly as in
     django.contrib.auth.decorators.user_passes_test
 - Add set_use_original_redirect_url_into_session function
   - Sets request.session["USE_ORIGINAL_REDIRECT_URL"] to True
 - Add is_request_user_unauthenticated function
   - Tests if request's user is unauthenticated i.e. not authenticated

### Backends: Set OIDC_OP_BASE_URL in (bf|ks|te)-pytest.yml 

Set OIDC_OP_BASE_URL to
https://tunnistus.test.hel.ninja/auth/.../openid-connect
in benefit, kesaseteli and tet backends' pytest GitHub Actions.

Rationale:
 - If OIDC_OP_BASE_URL is not set OIDC_OP_TOKEN_ENDPOINT will be set to
   "/token" which will be used in
   shared.azure_adfs.tests.test_adfs_login's test
   test_adfs_callback_original_redirect. This will call the following:
   - HelsinkiOIDCAuthenticationBackend.authenticate ->
     - OIDCAuthenticationBackend.get_token ->
       - requests.post ->
         - ... ->
           - PreparedRequest.prepare_url which will reject "/token" with
             MissingSchema exception: "Invalid URL '/token': No schema
             supplied. Perhaps you meant http:///token?".

Fixes test_adfs_callback_original_redirect test from
shared.azure_adfs.tests.test_adfs_login.

## Issues :bug:

YJDH-251

## Testing :alembic:

Steps:
 - Set up handler service and keep it running for the whole test:
   - `docker-compose -f docker-compose.handler.yml down`
   - `yarn handler --build`
 - Set `NEXT_PUBLIC_MOCK_FLAG=1` in `.env.kesaseteli`
 - `docker-compose -f docker-compose.youth.yml down`
 - `yarn youth --build`
 - Open `https://localhost:3100/` in web browser
   - Input youth application with valid data and submit it
     - You can generate valid Finnish personal identity codes using https://www.telepartikkeli.net/tunnusgeneraattori
     - Postcode must be 5 digits
     - You can use the same gmail.com email address multiple times by suffixing `+number` e.g. `x+1@gmail.com` is sent to the same address as `x@gmail.com`
   - Copy the youth application's ID from the browser's address bar somewhere to be used later, e.g.
     - `https://localhost:3100/fi/thankyou?id=7930a20a-3808-479f-8bf0-507f80c0c220` → `7930a20a-3808-479f-8bf0-507f80c0c220`
   - Activate the sent youth application by clicking on `AKTIVOI` link
 - `docker-compose -f docker-compose.youth.yml down`
 - Set `NEXT_PUBLIC_MOCK_FLAG=0` in `.env.kesaseteli`
 - `yarn youth`
 - Open `https://localhost:8000/v1/youthapplications/<id you copied earlier>/process/` e.g. `https://localhost:8000/v1/youthapplications/7930a20a-3808-479f-8bf0-507f80c0c220/process/` in web browser
   - If you get a page with HTTP ERROR 501 i.e. Not implemented everything's ok, you have permission to view the page but it has not yet been implemented
   - If you have logged into Helsinki city's AD and get redirected to https://localhost:3200/fi/403 with "Your connection is not private" or "Sinulla ei ole oikeuksia" text then try first logging out by clicking on `https://localhost:8000/oauth2/logout`, open Django admin page using link `https://localhost:8000/admin/`, log out from it, log into it with admin/admin, open users, open the user with the email address you used in logging into Helsinki city's AD, look at the user's groups, pick one of the ones with `adfs-` prefix and write down its suffix, e.g. `adfs-c5c7fb45-c112-4589-b599-e17fec5c14a3` → `c5c7fb45-c112-4589-b599-e17fec5c14a3`
   - Open `.env.kesaseteli` and put the value you just wrote down into it into the `ADFS_CONTROLLER_GROUP_UUIDS`, e.g. `ADFS_CONTROLLER_GROUP_UUIDS=c5c7fb45-c112-4589-b599-e17fec5c14a3`
   - `docker-compose -f docker-compose.youth.yml down`
   - `yarn youth`
   - Reopen the `https://localhost:8000/v1/youthapplications/<id you copied earlier>/process/` page and now you should get a page saying HTTP ERROR 501 i.e. Not implemented which is correct.

## Screenshots :camera_flash:

### Youth application processing page without handler permission:
![handler_403_page](https://user-images.githubusercontent.com/77663720/155475224-70044b86-696d-44af-96f1-13fe2f2889cc.png)

### Youth application processing page with handler permission:
![screenshot2](https://user-images.githubusercontent.com/77663720/154905624-bcb62cf9-f5fa-468b-bddf-ac2fea9b8e55.png)

☝️ This is ok, HTTP status 501 means not implemented which the processing functionality in actuality is.

## Additional notes :spiral_notepad:

nenonja commented:
> ☝️ This is ok, the redirected to page is coming in frontend PR https://github.com/City-of-Helsinki/yjdh/pull/814.

It is now merged so you can rebase or merge develop to get it.

karisal-anders commented:
> > ☝️ This is ok, the redirected to page is coming in frontend PR #814.
> 
> It is now merged so you can rebase or merge develop to get it.

Alright, thanks! Rebased and changed the testing instructions to accommodate the existence of the 403 handler page.

haavikko reviewed (approved):
LGTM!
